mlx/scheduler.h

```diff
@@ -35,8 +35,7 @@ struct StreamThread {
   }
 
   void thread_fn() {
```

Comment:
This needs the most scrutiny.  There were two problems with this code:

- both new_scoped_memory_pool and new_stream would try (and fail) to create autoreleasepools during static init time
- having an autoreleasepool scoped to the lifetime of thread_fn() means that the objects collected in there will be abandoned until thread_fn returns (likely at the termination of the process)

This moves the autoreleasepool scope to the level of `task()` so that the objects are collected when the work is done.  This also moves the initialization of `new_stream` to when the first task is ready -- this will be post static init.  This does change the timing (when) this work occurs, but the overall time of the processing should be the same (no additional work is done).